Can't stop with a savepoint？

**Describe the bug(Please use English)**
A clear and concise description of what the bug is.

**Environment :**
 - Flink version :  1.15.0
 - Flink CDC version: 2.2.1
 - Database and version: oracle 11g

**To Reproduce**
Steps to reproduce the behavior:
1. Thes test data :
flink  stop {jobId}
3. The test code :
SET parallelism.default = 4;
CREATE TABLE input(
    `L_DATE` DECIMAL(38, 0)
    ...
    ,PRIMARY KEY(`L_DATE`) NOT ENFORCED
) WITH (
    'connector' = 'oracle-cdc'
    ,'hostname' = '**'
    ,'port' = '**'
    ,'username' = 'flinkuser'
    ,'password' = '123456'
    ,'database-name' = 'orcl'
    ,'schema-name' = 'FLINKUSER'
    ,'table-name' = '**'
    ,'scan.startup.mode' = 'initial'
    ,'debezium.log.mining.strategy' = 'online_catalog'
    ,'debezium.log.mining.continuous.mine' = 'true'
    ,'debezium.database.tablename.case.insensitive' = 'false'
);
CREATE TABLE output(
    `L_DATE` BIGINT
    ...
) WITH (
    'connector' = 'kafka'
    ,'topic' = '**'
    ,'properties.bootstrap.servers' = ''
    ,'properties.group.id' = 'tmp_001'
    ,'scan.startup.mode' = 'earliest-offset'
    ,'format' = 'debezium-json'
);

SET execution.checkpointing.interval = 10s;
SET execution.checkpointing.tolerable-failed-checkpoints = 1000;
INSERT INTO output 
SELECT CAST(L_DATE AS BIGINT)
    ...
FROM input ;
5. The error :
Setting HADOOP_CONF_DIR=/etc/hadoop/conf because no HADOOP_CONF_DIR or HADOOP_CLASSPATH was set.
Setting HBASE_CONF_DIR=/etc/hbase/conf because no HBASE_CONF_DIR was set.
Suspending job "eb5ad9271f8a9824238845ff19ae171d" with a CANONICAL savepoint.

------------------------------------------------------------
 The program finished with the following exception:

org.apache.flink.util.FlinkException: Could not stop with a savepoint job "eb5ad9271f8a9824238845ff19ae171d".
	at org.apache.flink.client.cli.CliFrontend.lambda$stop$5(CliFrontend.java:588)
	at org.apache.flink.client.cli.CliFrontend.runClusterAction(CliFrontend.java:1026)
	at org.apache.flink.client.cli.CliFrontend.stop(CliFrontend.java:573)
	at org.apache.flink.client.cli.CliFrontend.parseAndRun(CliFrontend.java:1093)
	at org.apache.flink.client.cli.CliFrontend.lambda$main$10(CliFrontend.java:1156)
	at org.apache.flink.runtime.security.contexts.NoOpSecurityContext.runSecured(NoOpSecurityContext.java:28)
	at org.apache.flink.client.cli.CliFrontend.main(CliFrontend.java:1156)
Caused by: java.util.concurrent.ExecutionException: java.util.concurrent.CompletionException: org.apache.flink.runtime.checkpoint.CheckpointException: Task has failed.
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
	at org.apache.flink.client.cli.CliFrontend.lambda$stop$5(CliFrontend.java:586)
	... 6 more
Caused by: java.util.concurrent.CompletionException: org.apache.flink.runtime.checkpoint.CheckpointException: Task has failed.
	at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:326)
	at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:338)
	at java.util.concurrent.CompletableFuture.uniRelay(CompletableFuture.java:911)
	at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:899)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at org.apache.flink.runtime.rpc.akka.AkkaInvocationHandler.lambda$invokeRpc$1(AkkaInvocationHandler.java:252)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at org.apache.flink.util.concurrent.FutureUtils.doForward(FutureUtils.java:1387)
	at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.lambda$null$1(ClassLoadingUtils.java:93)
	at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:68)
	at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.lambda$guardCompletionWithContextClassLoader$2(ClassLoadingUtils.java:92)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at org.apache.flink.runtime.concurrent.akka.AkkaFutureUtils$1.onComplete(AkkaFutureUtils.java:45)
	at akka.dispatch.OnComplete.internal(Future.scala:299)
	at akka.dispatch.OnComplete.internal(Future.scala:297)
	at akka.dispatch.japi$CallbackBridge.apply(Future.scala:224)
	at akka.dispatch.japi$CallbackBridge.apply(Future.scala:221)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at org.apache.flink.runtime.concurrent.akka.AkkaFutureUtils$DirectExecutionContext.execute(AkkaFutureUtils.java:65)
	at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:68)
	at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1(Promise.scala:284)
	at scala.concurrent.impl.Promise$DefaultPromise.$anonfun$tryComplete$1$adapted(Promise.scala:284)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
	at akka.pattern.PromiseActorRef.$bang(AskSupport.scala:621)
	at akka.pattern.PipeToSupport$PipeableFuture$$anonfun$pipeTo$1.applyOrElse(PipeToSupport.scala:25)
	at akka.pattern.PipeToSupport$PipeableFuture$$anonfun$pipeTo$1.applyOrElse(PipeToSupport.scala:23)
	at scala.concurrent.Future.$anonfun$andThen$1(Future.scala:532)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:29)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:29)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: org.apache.flink.runtime.checkpoint.CheckpointException: Task has failed.
	at org.apache.flink.runtime.checkpoint.PendingCheckpoint.abort(PendingCheckpoint.java:545)
	at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.abortPendingCheckpoint(CheckpointCoordinator.java:2070)
	at org.apache.flink.runtime.checkpoint.CheckpointCoordinator.receiveDeclineMessage(CheckpointCoordinator.java:1038)
	at org.apache.flink.runtime.scheduler.ExecutionGraphHandler.lambda$declineCheckpoint$2(ExecutionGraphHandler.java:103)
	at org.apache.flink.runtime.scheduler.ExecutionGraphHandler.lambda$processCheckpointCoordinatorMessage$3(ExecutionGraphHandler.java:119)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.flink.runtime.checkpoint.CheckpointException: Task name with subtask : Source: input[1] -> Calc[2] (1/1)#0 Failure reason: Task has failed.
	at org.apache.flink.runtime.taskmanager.Task.declineCheckpoint(Task.java:1388)
	at org.apache.flink.runtime.taskmanager.Task.lambda$triggerCheckpointBarrier$3(Task.java:1331)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:822)
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:797)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$triggerCheckpointAsync$11(StreamTask.java:1094)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$SynchronizedStreamTaskActionExecutor.runThrowing(StreamTaskActionExecutor.java:93)
	at org.apache.flink.streaming.runtime.tasks.mailbox.Mail.run(Mail.java:90)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.processMailsWhenDefaultActionUnavailable(MailboxProcessor.java:338)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.processMail(MailboxProcessor.java:324)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:201)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:804)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:753)
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:948)
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:927)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:741)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:563)
	... 1 more
Caused by: java.util.concurrent.CompletionException: java.lang.Exception: Could not perform checkpoint 7 for operator Source: input[1] -> Calc[2] (1/1)#0.
	at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:326)
	at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:338)
	at java.util.concurrent.CompletableFuture.uniRelay(CompletableFuture.java:911)
	at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:899)
	... 15 more
Caused by: java.lang.Exception: Could not perform checkpoint 7 for operator Source: input[1] -> Calc[2] (1/1)#0.
	at org.apache.flink.streaming.runtime.tasks.StreamTask.triggerCheckpointAsyncInMailbox(StreamTask.java:1138)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$triggerCheckpointAsync$11(StreamTask.java:1085)
	... 12 more
Caused by: org.apache.flink.runtime.checkpoint.CheckpointException: Could not complete snapshot 7 for operator Source: input[1] -> Calc[2] (1/1)#0. Failure reason: Checkpoint was declined.
	at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.snapshotState(StreamOperatorStateHandler.java:269)
	at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.snapshotState(StreamOperatorStateHandler.java:173)
	at org.apache.flink.streaming.api.operators.AbstractStreamOperator.snapshotState(AbstractStreamOperator.java:348)
	at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.checkpointStreamOperator(RegularOperatorChain.java:227)
	at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.buildOperatorSnapshotFutures(RegularOperatorChain.java:212)
	at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.snapshotState(RegularOperatorChain.java:192)
	at org.apache.flink.streaming.runtime.tasks.SubtaskCheckpointCoordinatorImpl.takeSnapshotSync(SubtaskCheckpointCoordinatorImpl.java:647)
	at org.apache.flink.streaming.runtime.tasks.SubtaskCheckpointCoordinatorImpl.checkpointState(SubtaskCheckpointCoordinatorImpl.java:320)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$performCheckpoint$12(StreamTask.java:1253)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$SynchronizedStreamTaskActionExecutor.runThrowing(StreamTaskActionExecutor.java:93)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.performCheckpoint(StreamTask.java:1241)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.triggerCheckpointAsyncInMailbox(StreamTask.java:1126)
	... 13 more
Caused by: org.apache.flink.util.FlinkRuntimeException: Call snapshotState() on closed source, checkpoint failed.
	at com.ververica.cdc.debezium.DebeziumSourceFunction.snapshotState(DebeziumSourceFunction.java:309)
	at org.apache.flink.streaming.util.functions.StreamingFunctionUtils.trySnapshotFunctionState(StreamingFunctionUtils.java:118)
	at org.apache.flink.streaming.util.functions.StreamingFunctionUtils.snapshotFunctionState(StreamingFunctionUtils.java:99)
	at org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator.snapshotState(AbstractUdfStreamOperator.java:87)
	at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.snapshotState(StreamOperatorStateHandler.java:222)
	... 24 more

**Additional Description**
If applicable, add screenshots to help explain your problem.
OracleSorce And PgSource happen error, But MysqlSource is OK.

--- a/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/DebeziumSourceFunction.java
+++ b/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/DebeziumSourceFunction.java
@@ -71,6 +71,7 @@ import java.util.concurrent.Executors;
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.TimeUnit;
 
+import static com.ververica.cdc.debezium.internal.Handover.ClosedException.isGentlyClosedException;
 import static com.ververica.cdc.debezium.utils.DatabaseHistoryUtil.registerHistory;
 import static com.ververica.cdc.debezium.utils.DatabaseHistoryUtil.retrieveHistory;
 
@@ -304,9 +305,11 @@ public class DebeziumSourceFunction<T> extends RichSourceFunction<T>
     public void snapshotState(FunctionSnapshotContext functionSnapshotContext) throws Exception {
         if (handover.hasError()) {
             LOG.debug("snapshotState() called on closed source");
-            throw new FlinkRuntimeException(
-                    "Call snapshotState() on closed source, checkpoint failed.",
-                    handover.getError());
+            if (!isGentlyClosedException(handover.getError())) {
+                throw new FlinkRuntimeException(
+                        "Call snapshotState() on failed source, checkpoint failed.",
+                        handover.getError());
+            }
         } else {
             snapshotOffsetState(functionSnapshotContext.getCheckpointId());
             snapshotHistoryRecordsState();
@@ -590,5 +593,10 @@ public class DebeziumSourceFunction<T> extends RichSourceFunction<T>
     @VisibleForTesting
     public String getEngineInstanceName() {
         return engineInstanceName;
+    }
+
+    @VisibleForTesting
+    public Handover getHandover() {
+        return handover;
     }
 }

--- a/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/internal/Handover.java
+++ b/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/internal/Handover.java
@@ -219,11 +219,8 @@ public class Handover implements Closeable {
         }
 
         public static boolean isGentlyClosedException(Throwable cause) {
-            if (cause instanceof ClosedException
-                    && GENTLY_CLOSED_MESSAGE.equals(((ClosedException) cause).getMessage())) {
-                return true;
-            }
-            return false;
+            return cause instanceof ClosedException
+                    && GENTLY_CLOSED_MESSAGE.equals(((ClosedException) cause).getMessage());
         }
     }
 }

--- a/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/internal/Handover.java
+++ b/flink-connector-debezium/src/main/java/com/ververica/cdc/debezium/internal/Handover.java
@@ -191,6 +191,8 @@ public class Handover implements Closeable {
 
             if (error == null) {
                 error = new ClosedException();
+            } else if (!(error instanceof ClosedException)) {
+                error = new ClosedException("Close handover with error.", error);
             }
             lock.notifyAll();
         }
@@ -205,5 +207,23 @@ public class Handover implements Closeable {
     public static final class ClosedException extends Exception {
 
         private static final long serialVersionUID = 1L;
+
+        private static final String GENTLY_CLOSED_MESSAGE = "Close handover gently.";
+
+        public ClosedException() {
+            super(GENTLY_CLOSED_MESSAGE);
+        }
+
+        public ClosedException(String message, Throwable cause) {
+            super(message, cause);
+        }
+
+        public static boolean isGentlyClosedException(Throwable cause) {
+            if (cause instanceof ClosedException
+                    && GENTLY_CLOSED_MESSAGE.equals(((ClosedException) cause).getMessage())) {
+                return true;
+            }
+            return false;
+        }
     }
 }

--- a/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/LegacyMySqlSourceTest.java
+++ b/flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/LegacyMySqlSourceTest.java
@@ -19,6 +19,7 @@ package com.ververica.cdc.connectors.mysql;
 import org.apache.flink.api.java.tuple.Tuple2;
 import org.apache.flink.core.testutils.CheckedThread;
 import org.apache.flink.runtime.state.StateSnapshotContextSynchronousImpl;
+import org.apache.flink.util.FlinkRuntimeException;
 
 import com.fasterxml.jackson.core.JsonParseException;
 import com.jayway.jsonpath.JsonPath;
@@ -29,6 +30,8 @@ import com.ververica.cdc.connectors.mysql.testutils.UniqueDatabase;
 import com.ververica.cdc.connectors.utils.TestSourceContext;
 import com.ververica.cdc.debezium.DebeziumSourceFunction;
 import com.ververica.cdc.debezium.history.FlinkJsonTableChangeSerializer;
+import com.ververica.cdc.debezium.internal.Handover;
+import io.debezium.DebeziumException;
 import io.debezium.document.Document;
 import io.debezium.document.DocumentWriter;
 import io.debezium.relational.Column;
@@ -907,6 +910,143 @@ public class LegacyMySqlSourceTest extends LegacyMySqlTestBase {
                                                     "Retrieve schema history failed, the schema records for engine %s has been removed,"
                                                             + " this might because the debezium engine has been shutdown due to other errors.",
                                                     engineInstanceName)));
+                } finally {
+                    source.close();
+                    runThread.sync();
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testSnapshotOnClosedSource() throws Exception {
+        final TestingListState<byte[]> offsetState = new TestingListState<>();
+        final TestingListState<String> historyState = new TestingListState<>();
+
+        {
+            try (Connection connection = database.getJdbcConnection();
+                    Statement statement = connection.createStatement()) {
+                // Step-1: start the source from empty state
+                final DebeziumSourceFunction<SourceRecord> source = createMySqlBinlogSource();
+                final TestSourceContext<SourceRecord> sourceContext = new TestSourceContext<>();
+                // setup source with empty state
+                setupSource(source, false, offsetState, historyState, true, 0, 1);
+
+                final CheckedThread runThread =
+                        new CheckedThread() {
+                            @Override
+                            public void go() throws Exception {
+                                source.run(sourceContext);
+                            }
+                        };
+                runThread.start();
+
+                // wait until the source finishes the database snapshot
+                List<SourceRecord> records = drain(sourceContext, 9);
+                assertEquals(9, records.size());
+
+                // state is still empty
+                assertEquals(0, offsetState.list.size());
+                assertEquals(0, historyState.list.size());
+
+                statement.execute(
+                        "INSERT INTO `products` VALUES (110,'robot','Toy robot',1.304)"); // 110
+
+                int received = drain(sourceContext, 1).size();
+                assertEquals(1, received);
+
+                // Step-2: trigger a checkpoint
+                synchronized (sourceContext.getCheckpointLock()) {
+                    // trigger checkpoint-1
+                    source.snapshotState(new StateSnapshotContextSynchronousImpl(101, 101));
+                }
+
+                assertTrue(historyState.list.size() > 0);
+                assertTrue(offsetState.list.size() > 0);
+
+                // Step-3: mock the engine stop with savepoint, trigger a
+                // checkpoint on closed source
+                final Handover handover = source.getHandover();
+                handover.close();
+                synchronized (sourceContext.getCheckpointLock()) {
+                    // trigger checkpoint-2
+                    source.snapshotState(new StateSnapshotContextSynchronousImpl(102, 102));
+                }
+
+                assertTrue(historyState.list.size() > 0);
+                assertTrue(offsetState.list.size() > 0);
+
+                source.close();
+                runThread.sync();
+            }
+        }
+    }
+
+    @Test
+    public void testSnapshotOnFailedSource() throws Exception {
+        final TestingListState<byte[]> offsetState = new TestingListState<>();
+        final TestingListState<String> historyState = new TestingListState<>();
+
+        {
+            try (Connection connection = database.getJdbcConnection();
+                    Statement statement = connection.createStatement()) {
+                // Step-1: start the source from empty state
+                final DebeziumSourceFunction<SourceRecord> source = createMySqlBinlogSource();
+                final TestSourceContext<SourceRecord> sourceContext = new TestSourceContext<>();
+                // setup source with empty state
+                setupSource(source, false, offsetState, historyState, true, 0, 1);
+
+                final CheckedThread runThread =
+                        new CheckedThread() {
+                            @Override
+                            public void go() throws Exception {
+                                source.run(sourceContext);
+                            }
+                        };
+                runThread.start();
+
+                // wait until the source finishes the database snapshot
+                List<SourceRecord> records = drain(sourceContext, 9);
+                assertEquals(9, records.size());
+
+                // state is still empty
+                assertEquals(0, offsetState.list.size());
+                assertEquals(0, historyState.list.size());
+
+                statement.execute(
+                        "INSERT INTO `products` VALUES (110,'robot','Toy robot',1.304)"); // 110
+
+                int received = drain(sourceContext, 1).size();
+                assertEquals(1, received);
+
+                // Step-2: trigger a checkpoint
+                synchronized (sourceContext.getCheckpointLock()) {
+                    // trigger checkpoint-1
+                    source.snapshotState(new StateSnapshotContextSynchronousImpl(101, 101));
+                }
+
+                assertTrue(historyState.list.size() > 0);
+                assertTrue(offsetState.list.size() > 0);
+
+                // Step-3: mock the engine stop due to underlying debezium exception, trigger a
+                // checkpoint on failed source
+                final Handover handover = source.getHandover();
+                handover.reportError(new DebeziumException("Mocked debezium exception"));
+                handover.close();
+                try {
+                    synchronized (sourceContext.getCheckpointLock()) {
+                        // trigger checkpoint-2
+                        source.snapshotState(new StateSnapshotContextSynchronousImpl(102, 102));
+                    }
+                    fail("Should fail.");
+                } catch (Exception e) {
+                    assertTrue(e instanceof FlinkRuntimeException);
+                    assertTrue(
+                            e.getMessage()
+                                    .contains(
+                                            "Call snapshotState() on failed source, checkpoint failed."));
+                    assertTrue(e.getCause() instanceof Handover.ClosedException);
+                    assertTrue(e.getCause().getMessage().contains("Close handover with error."));
                 } finally {
                     source.close();
                     runThread.sync();

